Sources/MongoSwift/ChangeStream.swift

```diff
@@ -24,35 +27,39 @@ public struct ResumeToken: Codable, Equatable {
 /// A MongoDB ChangeStream.
 /// - SeeAlso: https://docs.mongodb.com/manual/changeStreams/
 public class ChangeStream<T: Codable>: Sequence, IteratorProtocol {
-    /// A `ResumeToken` associated with the most recent event seen by the change stream.
-    public private(set) var resumeToken: ResumeToken?
-
-    /// A `MongoClient` stored to make sure the source client stays valid until the change stream is destroyed.
-    private let client: MongoClient
-
-    /// A `Connection` stored to make sure the client connection stays valid until the change stream is destroyed.
-    private let connection: Connection
+    /// Enum for tracking the state of a change stream.
+    internal enum State {
+        /// Indicates that the change stream is still open. Stores a pointer to the `mongoc_change_stream_t`, along
+        /// with the source connection, client, and possibly session to ensure they are kept alive as long
+        /// as the change stream is.
+        case open(changeStream: OpaquePointer, connection: Connection, client: MongoClient, session: ClientSession?)
+        case closed
+    }
 
-    /// A `ClientSession` stored to make sure the session stays valid until the change stream is destroyed.
-    private let session: ClientSession?
+    /// The state of this change stream.
+    internal private(set) var state: State
 
-    /// A reference to the `mongoc_change_stream_t` pointer.
-    private let changeStream: OpaquePointer
+    /// A `ResumeToken` associated with the most recent event seen by the change stream.
+    public internal(set) var resumeToken: ResumeToken?
 
     /// Decoder for decoding documents into type `T`.
-    private let decoder: BSONDecoder
+    internal let decoder: BSONDecoder
 
-    /// Used for storing Swift errors.
-    private var swiftError: Error?
+    /// The error that occurred while iterating the change stream, if one exists. This should be used to check
+    /// for errors after `next()` returns `nil`.
+    public private(set) var error: Error?
 
     /**
-     * The error that occurred while iterating the change stream, if one exists. This should be used to check
-     * for errors after `next()` returns `nil`.
+     * Retrieves any error that occured in mongoc or on the server while iterating the change stream. Returns nil if
+     * this change stream is already closed, or if no error occurred.
      *  - Errors:
      *    - `DecodingError` if an error occurs while decoding the server's response.
      */
-    public var error: Error? {
-        if let err = self.swiftError {
+    private func getChangeStreamError() -> Error? {
+        guard case let .open(changeStream, _, _, _) = self.state else {
+            return nil
+        }
+        if let err = self.error {
```

Comment:
I think we can skip this check because this could return any error, not just an error from mongoc (e.g. some decoding error).
I also don't think there's currently a code path where we call this method and this property would have already been set to non-nil